### PR TITLE
Implement Reddit liquid tag

### LINKED ISSUE #1536: Reddit Liquid Tag
**Describe the solution you'd like**
It would be nice to have a liquid tag to integrate reddit submissions

**Describe alternatives you've considered**
I haven't researched intensively but found that Reddit has an embedding mechanism (oEmbed ?)

**Additional context**
[A maybe useful GH wiki page about Reddit embedding.](https://github.com/reddit-archive/reddit/wiki/oEmbed)

### PR DESCRIPTION
**Pull request by cesc1989**
## What type of PR is this? (check all applicable)
- [x] Feature

## Description
Implement Liquid tag for reddit content.

It takes the post URL and renders a block of content based on the JSON response of requesting that URL and appending `.json`:

```
url =  https://www.reddit.com/r/IAmA/comments/afvl2w/im_scott_from_scotts_cheap_flights_my_profession

json_url =  https://www.reddit.com/r/IAmA/comments/afvl2w/im_scott_from_scotts_cheap_flights_my_profession.json
```

I tried to emulate the same way the Medium liquid tag was coded and that's why the request is made and handled in a different class(`RedditJsonFromUrlService`)

I'm still pending to add tests and improve the styling of the reddit content.

## Related Tickets & Documents
This PR resolves #1536 

## Added to documentation?

- [x] docs.dev.to

## What gif best describes this PR or how it makes you feel?

grrrreat

### CODE REVIEW

**Comment by benhalpern** [comment]
Looking good so far, thanks for this

**Comment by cesc1989** [comment]
Hi @benhalpern 👋 

I'm close to finish this but got a couple of questions. Hope this is the place to ask.

1. When rendering reddit content that has image, this happens:
01-reddit-image

Don't know where in the codebase happens but seems the image URL is wrapped to make it lazy load. I'm not very sure how to proceed fix/proceed/handle this situation and would appreciate your input.

2. Reddit content again but with text only:
02-reddit-text
I tried to parse the markdown content but couldn't successfully do it. Again, what would be a way to solve this _in_ Dev?

3. Final question is this in the PR checks:

> codeclimate/diff-coverage — 17% (50% threshold)

Don't understand what's going on there. I installed the codeclimate firefox plugin but can't see the diff.

I appreciate your time 😄

**Comment by benhalpern** [comment]
@cesc1989 I believe it is likely the images are only failing locally because you don't have Cloudinary fully set up. We can confirm this, but that is my suspicion right now.

With regards to the text/markdown/html, we should be able to parse as markdown in our normal parser.

CodeClimate is down because it needs tests. Check out `spec/liquid_tags` and add tests similar to what's in there for similar tags.

@maestromac @Zhao-Andy if either of you want to add some guidance or suggestions here when you could, that would be great.

**Comment by maestromac** [comment]
Hello @cesc1989 
1. We lazy load all images and defer the image to Cloudinary. I've tested your branch locally and it seems to be loading the image. Is it not loading at all on your end?

2. We use `Redcarpet` and you can try to play around with it to see how it renders your Reddit's content (ie `markdown = Redcarpet::Markdown.new(renderer, extensions = {})` or take a look to see how we use it [here](https://github.com/thepracticaldev/dev.to/blob/master/app/labor/markdown_parser.rb#L12)). If that doesn't work and complicate things, would it be possible to just render the plain text without the markdown?

3. Seems like `diff-coverage` is no longer the issue here

**Comment by benhalpern** [comment]
> We lazy load all images and defer the image to Cloudinary. I've tested your branch locally and it seems to be loading the image. Is it not loading at all on your end?

@maestromac I believe this is because the cloudinary key is set to `Optional` in their env. So it probably works for you and not them if they don't have it set properly. I'd say this part of the PR should be good to go.

We mostly just need to proper markdown parsing now.

**Comment by cesc1989** [comment]
Thanks a bunch! @benhalpern @maestromac 

I finally could render the text content properly 😄 I think the PR is ready to be reviewed/merged. Let me know if there's anything to change/improve.

**Comment by Zhao-Andy** [comment]
Hey @cesc1989 sorry for the late review! I'm reviewing it now and noticed the icon could be more true to Reddit's icon. Do you mind if I push up a new SVG file?

Have some additional comments to make, but overall looks great!

File: app/views/liquids/_reddit.html.erb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<!-- app/views/liquids/_reddit.html.erb -->
```
Comment (Zhao-Andy) [inline app/views/liquids/_reddit.html.erb]: We generally don't have comments about the file's path, and we could have it but that should probably be in a different PR. Also since this is an HTML comment let's remove this one at least. :)

File: app/views/liquids/_reddit.html.erb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      </div>
+    </header>
+  </div>
+  <div class="ltag__reddit--body">
+    <p>
+      <% if post_hint == 'self' %>
+        <%= markdown_text %>
+      <% elsif post_hint == 'image' %>
+        <img src="<%= thumbnail %>" alt="<%= title %>">
+      <% end %>
```
Comment (Zhao-Andy) [inline app/views/liquids/_reddit.html.erb]: I think this might be better if it's:
```ruby
if post_hint == "image"
 <img...>
else
  <%= markdown_text %>
end
```

because I ran into an edge case where `post_hint` is `nil`. For example:
https://www.reddit.com/r/Android/comments/dvzyeb/unpopular_opinion_3button_navigation_is_just_flat.json

File: docs/frontend/liquid-tags.md [hunk 3]
```diff
@@ -41,4 +41,5 @@ Here is a bunch of liquid tags supported on DEV:
 {% spotify spotify:episode:5V4XZWqZQJvbddd31n56mf %}
 {% blogcast 1234 %}
 {% kotlin https://pl.kotl.in/owreUFFUG %}
+{% reddit https://www.reddit.com/r/aww/comments/ag3s4b/ive_waited_28_years_to_finally_havr_my_first_pet %}
```
Comment (Zhao-Andy) [inline docs/frontend/liquid-tags.md]: Thanks for adding this to the docs! Could you also add an example to the editor guide? You can find the partial file in `app/views/pages/_editor_guide_text.html.erb`.

File: spec/liquid_tags/reddit_tag_spec.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# spec/liquid_tags/reddit_tag_spec.rb
```
Comment (Zhao-Andy) [inline spec/liquid_tags/reddit_tag_spec.rb]: Let's remove this comment as per a previous review comment.

File: app/services/reddit_json_from_url_service.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
+# app/services/reddit_json_from_url_service.rb
```
Comment (Zhao-Andy) [inline app/services/reddit_json_from_url_service.rb]: Let's remove this comment as per a previous review comment.

File: app/liquid_tags/reddit_tag.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+# app/liquid_tags/reddit_tag.rb
```
Comment (Zhao-Andy) [inline app/liquid_tags/reddit_tag.rb]: Let's remove this comment as per a previous review comment.

File: app/assets/images/reddit-icon.svg [hunk 7]
```diff
@@ -0,0 +1,63 @@
+<?xml version="1.0" encoding="iso-8859-1"?>
```
Comment (Zhao-Andy) [inline app/assets/images/reddit-icon.svg]: Leaving a reminder about changing this SVG. If you'd like, I can push up the new SVG, or you can copy and paste the following SVG code:
```
<svg aria-hidden="true" focusable="false" data-prefix="fab" data-icon="reddit" class="svg-inline--fa fa-reddit fa-w-16" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="#FE5022" d="M201.5 305.5c-13.8 0-24.9-11.1-24.9-24.6 0-13.8 11.1-24.9 24.9-24.9 13.6 0 24.6 11.1 24.6 24.9 0 13.6-11.1 24.6-24.6 24.6zM504 256c0 137-111 248-248 248S8 393 8 256 119 8 256 8s248 111 248 248zm-132.3-41.2c-9.4 0-17.7 3.9-23.8 10-22.4-15.5-52.6-25.5-86.1-26.6l17.4-78.3 55.4 12.5c0 13.6 11.1 24.6 24.6 24.6 13.8 0 24.9-11.3 24.9-24.9s-11.1-24.9-24.9-24.9c-9.7 0-18 5.8-22.1 13.8l-61.2-13.6c-3-.8-6.1 1.4-6.9 4.4l-19.1 86.4c-33.2 1.4-63.1 11.3-85.5 26.8-6.1-6.4-14.7-10.2-24.1-10.2-34.9 0-46.3 46.9-14.4 62.8-1.1 5-1.7 10.2-1.7 15.5 0 52.6 59.2 95.2 132 95.2 73.1 0 132.3-42.6 132.3-95.2 0-5.3-.6-10.8-1.9-15.8 31.3-16 19.8-62.5-14.9-62.5zM302.8 331c-18.2 18.2-76.1 17.9-93.6 0-2.2-2.2-6.1-2.2-8.3 0-2.5 2.5-2.5 6.4 0 8.6 22.8 22.8 87.3 22.8 110.2 0 2.5-2.2 2.5-6.1 0-8.6-2.2-2.2-6.1-2.2-8.3 0zm7.7-75c-13.6 0-24.6 11.1-24.6 24.9 0 13.6 11.1 24.6 24.6 24.6 13.8 0 24.9-11.1 24.9-24.6 0-13.8-11-24.9-24.9-24.9z"></path></svg>
```

If you do want to do it yourself, remember to update the VCR cassette test, too. :)

File: docs/frontend/liquid-tags.md (hunk 3 above)
Comment (cesc1989) [reply docs/frontend/liquid-tags.md]: Yes. In fact, I was expecting this to be mentioned before adding another commit to the PR xD

I'll add the example

File: app/assets/images/reddit-icon.svg (hunk 7 above)
Comment (cesc1989) [reply app/assets/images/reddit-icon.svg]: I can do it 😄

**Comment by cesc1989** [comment]
@Zhao-Andy just submitted the suggested changes 😄 

The reddit icon does look nicer. Thanks!

**Comment by cesc1989** [comment]
BTW, @Zhao-Andy I think the code in `app/views/pages/_editor_guide_text.html.erb` would benefit from using **more** partials(maybe per tag). I freaked out when saw all that HTML with no order whatsoever 😅 

How do you handle this kind of refactors?

**Comment by cesc1989** [comment]
> Actually the size of the VCR is concerning. I need to think about a better way to test this without relying on VCR.

@maestromac thanks for reviewing!

I'll take a look at it tomorrow and see how can it be improved.

**Comment by mstruve** [comment]
Hey @cesc1989! 

Thank you so much for the PR to add this incredible feature! With that said, we are in the process of going through and cleaning up some of our PR history to help us better prioritize what we need to be focusing on. Since this PR has not been active for a while and has some conflicts I am going to close it for now. If you would still like to get this merged please feel free to resolve the conflicts, make the requested changes, and then open the PR back up. We would be more than happy to help push this through if you would like!

Thanks so much for being an awesome contributor! Ping us anytime if you have questions!

**Comment by cesc1989** [comment]
Thank you @mstruve for the heads up.

I have this inline to work but had it completely forgotten. Soon I'll be tackling the conflicts and the last reviews 🤠 👍

**Comment by cesc1989** [comment]
Hi @mstruve @Zhao-Andy 

I just pushed conflicts solved and the two feedback items provided. Do I have to open another PR?

**Comment by rhymes** [comment]
Could you rebase this on top of master and force push? Then you can reopen this PR I think. Thanks @cesc1989!
